### PR TITLE
Regressor fixed

### PR DESCRIPTION
**Pull request by jindrahelcl**
replaces #412

### CODE REVIEW

File: neuralmonkey/runners/regression_runner.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class RegressionRunner(BaseRunner):
+
+    def __init__(self,
+                 output_series: str,
+                 decoder: Any,
+                 postprocess: Callable[[float], float]=None) -> None:
+        super(RegressionRunner, self).__init__(output_series, decoder)
+
```
Comment (jlibovicky) [inline neuralmonkey/runners/regression_runner.py]: Kód volaný z konfiguráku by měl mít typeguard.

File: neuralmonkey/runners/regression_runner.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+from neuralmonkey.runners.base_runner import (BaseRunner, Executable,
+                                              ExecutionResult, NextExecute)
+# pylint: disable=too-few-public-methods
+
+
+class RegressionRunner(BaseRunner):
+
+    def __init__(self,
+                 output_series: str,
+                 decoder: Any,
```
Comment (jlibovicky) [inline neuralmonkey/runners/regression_runner.py]: Decoder v tomhle případě musí být ten `SequenceRegresson`, nebo ne? I kdyby ne, tak by to měl být aspoň `ModelPart`.

File: neuralmonkey/decoders/sequence_regressor.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                "learning_step", [], trainable=False,
+                initializer=tf.constant_initializer(0))
+
+            self.dropout_placeholder = \
+                tf.placeholder(tf.float32, name="dropout_plc")
+            self.gt_inputs = tf.placeholder(tf.float32, shape=[None],
+                                            name="targets")
+
+            mlp_input = tf.concat([enc.encoded for enc in encoders], 1)
+            mlp = MultilayerPerceptron(
```
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Tady nemá být `nn.mlp.MultilLayerPerceptron`, což je klasifikátor, ale jenom `nn.projection.multilayerprojection`. Ten MLP dělá navíc softmax, který v tomhle případě vůbec není potřeba.

File: neuralmonkey/decoders/sequence_regressor.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                mlp_input, layers, self.dropout_placeholder, 1,
+                activation_fn=self.activation_fn)
+
+            assert_shape(mlp.logits, [-1, 1])
+
+            self.decoded_logit = mlp.logits
```
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Logit je energy softmaxu před normalizací. Tomuhle bych říkal něco jako `predicted`.

File: neuralmonkey/decoders/sequence_regressor.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        self.max_output_len = 1
+
+        with tf.variable_scope(name):
+            self.learning_step = tf.get_variable(
+                "learning_step", [], trainable=False,
+                initializer=tf.constant_initializer(0))
+
+            self.dropout_placeholder = \
+                tf.placeholder(tf.float32, name="dropout_plc")
+            self.gt_inputs = tf.placeholder(tf.float32, shape=[None],
```
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Asi není potřeba, aby ten regresor dával na výstup jenom jedno číslo, klidně může mít vektor.

File: neuralmonkey/decoders/sequence_regressor.py (hunk 4 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: puvodne tam bylo predicted, problem je ze runnery a dalsi skripty od dekoderu predpokladaji presne tuhle informaci, proto jsem to prejmenoval na decoded_logit. Tento problem by mel vyresit abstraktni rodic decoderu ktery se planuje

File: neuralmonkey/runners/regression_runner.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+            predictions_sum += sess_result["prediction"]
+
+        predictions = predictions_sum / len(results)
+
+        if self._postprocess is not None:
+            predictions = self._postprocess(predictions)
+
+        self.result = ExecutionResult(
+            outputs=predictions.tolist(),
```
Comment (tomkocmi) [inline neuralmonkey/runners/regression_runner.py]: tohle s tim .tolist() jsem sem pridal aby to vypisovalo validaci. Ted jsem si ale vsiml ze testovani na tomhle pada pri zapisu do vystupniho souboru na radce 394 learning_utils. Reseni by bylo prevest predictions do listu stringu aby se to jak vypisovalo, tak pak fungovalo pri testovani.

File: neuralmonkey/decoders/sequence_regressor.py (hunk 5 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: na jednu stranu souhlasim, ale myslim, ze to rozbije hrozne veci okolo (napriklad co jsem komentoval ten problem s vypisovanim ve validaci)

File: neuralmonkey/decoders/sequence_regressor.py (hunk 4 above)
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Kde se vyžaduje, aby to mělo `decoded_logit`? Tohle se stejně bude pouštět jenom `RegressionRunnerem` a můžeš napsat, co chceš. Trainer  si z toho regresoru vezme jenom `cost`. S ničím jiným by ten regresor neměl interagovat.

File: neuralmonkey/runners/regression_runner.py (hunk 6 above)
Comment (jlibovicky) [inline neuralmonkey/runners/regression_runner.py]: Regresor by IMHO měl vracet list floatů. Pokud se někde neumí vytisknout `float`, tak tam připiš, jak se vytiskne `float`. Umět vypsat float je obecně užitečná vlastnost.

File: neuralmonkey/decoders/sequence_regressor.py (hunk 5 above)
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Vrátilo by to list numpy arrayů. Za ty se ve validaci tiskne zástupný string `np.array` při výpisu testovacích dat by se tam dala podmínka, že se vytiskne `str(np.array)`, to by nebyl problém.

Ale jestli se ti nechce dělat, tak to samozřejmě nedělej, udělá se to, až to někdo bude potřebovat.

File: neuralmonkey/decoders/sequence_regressor.py (hunk 3 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: kdyz se specificky nevola, tak softmax nedela.

Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: kazdopadne trenoval jsem to ted na datasetu s rozmezim <1,5> a nenaucilo se to outputovat do tohodle rozpeti, strilelo to i cisla <1 a nikdy nic >2

Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: To, že se nevolá ten softmax nemění nic na sémantice těch funkcí a tříd. `MultiLayerPerceptron` je klasifikátor a tady se nepoužívá klasifikátor ale projekce. Podle mě to zbytečně mate čtenáře kódu, který vidí perceptron, ale ona je to jenom projekce.

Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Outputovalo to i záporná čísla?

Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: nn pouze rozmezi (0,2) ... koukal jsem ted do logu a vidim tam cisla pouze v rozmezi (0.6,1.3)

File: neuralmonkey/runners/regression_runner.py [hunk 7]
```diff
@@ -15,22 +17,23 @@ class RegressionRunner(BaseRunner):
...
+                       compute_losses: bool = False,
                        summaries=True) -> Executable:
 
         if compute_losses:
             fetches = {"mse": self._decoder.cost}
         else:
             fetches = {}
 
-        fetches["prediction"] = self._decoder.decoded_logit
+        fetches["prediction"] = self._decoder.predicted
```
Comment (jindrahelcl) [inline neuralmonkey/runners/regression_runner.py]: Asi bych to udělal tak, že sem bych plácnul `self._decoder.decoded` a `predicted` vlastnost sequence regressoru udělal protected (`_predicted`).

Comment (jlibovicky) [inline neuralmonkey/runners/regression_runner.py]: Mně to přijde dobrý takhle. Za něčím, co by se jmenovalo `decoded` bych čekal něco disktrétního, ne float.

File: neuralmonkey/decoders/sequence_regressor.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                                            name="targets")
+
+            mlp_input = tf.concat([enc.encoded for enc in encoders], 1)
+            # TODO extend it to output into multidimensional space
+            layers.append(1)
```
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: tohle nebude fungovat protoze mlp pridava na konec relu. Vratil bych sem MultilayerPerceptron protoze ten dela presne to co chceme (udela vrstvy, spravne jim da dropout a hlavne posledni vrstvu neudela ReLu na posledni vrstve) takze se to nemusi manualne vytvaret v regressoru.

File: neuralmonkey/runners/regression_runner.py (hunk 7 above)
Comment (jindrahelcl) [inline neuralmonkey/runners/regression_runner.py]: hmm. ale to decoded je tam jako nějaká abstraktní vlastnost dekodéru.. Možná by tedy šlo vyhodit decoded, nevim, kde se to ještě používá.

Comment (jlibovicky) [inline neuralmonkey/runners/regression_runner.py]: Nikde nemáme definované abstrakní vlastosti dekodéru, jediný společný předek je `ModelPart`. Pokud nějaký runner (třeba GreedyRunner nebo LabelRunner) spoléhají na `decoded`, tak to používají jako index do slovníku.

File: neuralmonkey/decoders/sequence_regressor.py (hunk 8 above)
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_regressor.py]: Radšich bych přidal vstvu navíc tady. MLP je klasifikátor, logity jsou nenormalizovaná distribuce, až ten kód někdo bude číst, tak ho to bude zbytečně mást.

Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_regressor.py]: No nelibi se mi to. Jinak stale mi ten regresor nefunguje vypada to jako kdyz by se nekde po ceste neco prumerovalo pres vsecny batche (vraci to podobne vysledky pro vsechny data ... defakto to najde idealni hodnotu jak bez znalosti vstupu mit nejmensi chybu a nevim kde je problem)
